### PR TITLE
Implement Okapi BM25 variants in Gensim

### PR DESCRIPTION
**Pull request by Witiko**
This pull request implements the `gensim.models.bm25model` module, which contains an implementation of [the Okapi BM25 model][1] and its modifications (Lucene BM25 and ATIRE) as discussed in https://github.com/RaRe-Technologies/gensim/issues/2592#issuecomment-866799145. The module acts as a replacement for the `gensim.summarization.bm25model` module deprecated and removed in Gensim 4. The module should supersede the `gensim.models.tfidfmodel` module as the baseline weighting function for information retrieval and related NLP tasks.

Most implementations of BM25 such as [the rank-bm25 library][2] combine indexing with weighting and often forgo dictionary building for a speed improvement at indexing time (but hefty penalty at retrieval time). To give an example, here is how a user would search for documents with rank-bm25:

``` python
>>> from rank_bm25 import BM25Okapi
>>>
>>> corpus = [["Hello", "world"], ["bar", "bar"], ["foo", "bar"]]
>>> bm25_model = BM25Okapi(corpus)
>>>
>>> query = ["Hello", "bar"]
>>> similarities = bm25_model.get_scores(query)
>>> similarities

array([0.51082562, 0.09121886, 0.0638532 ])

>>> best_document, = bm25_model.get_top_n(query, corpus, n=1)
>>> best_document

['Hello', 'world']
```

As you can see, the interface is convenient, but retrieval is slow due to the lack of a dictionary. Furthermore, any advanced operations such as pruning the dictionary, applying semantic matching (e.g. SCM) and query expansion (e.g. RM3), or sharding the index are unavailable.

By contrast, the `gensim.models.bm25` module separates the three operations. To give an example, here is how a user would search for documents with the `gensim.models.bm25` module:

``` python
>>> from gensim.corpora import Dictionary
>>> from gensim.models import TfidfModel, OkapiBM25Model
>>> from gensim.similarities import SparseMatrixSimilarity
>>> import numpy as np
>>>
>>> corpus = [["Hello", "world"], ["bar", "bar"], ["foo", "bar"]]
>>> dictionary = Dictionary(corpus)
>>> bm25_model = OkapiBM25Model(dictionary=dictionary)
>>> bm25_corpus = bm25_model[list(map(dictionary.doc2bow, corpus))]
>>> bm25_index = SparseMatrixSimilarity(bm25_corpus, num_docs=len(corpus), num_terms=len(dictionary),
...                                    normalize_queries=False, normalize_documents=False)
>>>
>>> query = ["Hello", "bar"]
>>> tfidf_model = TfidfModel(dictionary=dictionary, smartirs='bnn')  # Enforce binary weighting of queries
>>> tfidf_query = tfidf_model[dictionary.doc2bow(query)]
>>>
>>> similarities = bm25_index[tfidf_query]
>>> similarities

array([0.51082563, 0.09121886, 0.0638532 ], dtype=float32)

>>> best_document = corpus[np.argmax(similarities)]
>>> best_document

['Hello', 'world']
```

Tasks:
- [x] Add Okapi BM25, ~~BM25L and BM25<sup>+</sup>~~ [[1][10], [2][11]], Lucene BM25 [[3][8], [4][9]], and ATIRE BM25 [[3][8], [5][6]].
- [x] Add comments and docstrings to `models.bm25`.
- [x] Add comments and docstrings to `similarities.docsim`.
- [x] Add BM25 to [the `run_topics_and_transformations` autoexample][7].
- [x] Add `normalize_queries=True, normalize_documents=True` named parameters to `SparseMatrixSimilarity`, `DenseMatrixSimilarity`, and `SoftCosineSimilarity` classes as discussed in https://github.com/RaRe-Technologies/gensim/pull/3304#issuecomment-1061031969 and on [the Gensim mailing list][4]. Deprecate the `normalize` named parameter of `SoftCosineSimilarity`. Add `normalize_queries=False, normalize_documents=False` to TF-IDF and BM25 examples.

 [1]: https://en.wikipedia.org/wiki/Okapi_BM25
 [2]: https://pypi.org/project/rank-bm25/
 [3]: https://radimrehurek.com/gensim/auto_examples/core/run_topics_and_transformations.html
 [4]: https://groups.google.com/g/gensim/c/wzQ1cIl9Img
 [5]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7148026/
 [6]: http://www.cs.otago.ac.nz/homepages/andrew/papers/2014-2.pdf
 [7]: https://radimrehurek.com/gensim/auto_examples/core/run_topics_and_transformations.html
 [8]: https://link.springer.com/chapter/10.1007/978-3-030-45442-5_4/tables/1
 [9]: https://speakerdeck.com/elastic/improved-text-scoring-with-bm25?slide=3
 [10]: https://github.com/RaRe-Technologies/gensim/pull/3304#issuecomment-1069500728
 [11]: https://github.com/RaRe-Technologies/gensim/pull/3304#issuecomment-1085783628

### CODE REVIEW

**Comment by piskvorky** [comment]
Pretty nice! We'll look into this after the 4.2 release.

**Comment by Witiko** [comment]
@piskvorky Thanks. No need to look just yet, I have tried some benchmarks and the code still seems to have issues both with respect to speed and correctness. I will let you know when the PR is ready for review.

**Comment by Witiko** [comment]
I have experimentally confirmed compatibility of 4b50675 with [the rank-bm25 library][1]:

- [Information retrieval on TREC 6–8 collection with rank-bm25: 19.54% MAP score][3]
- [Information retrieval on TREC 6–8 collection with Gensim: 19.54% MAP score][4]

I have also [outlined some issues][2] with the default behavior of the `DenseMatrixSimilarity` and `SparseMatrixSimilarity` indexes, which are likely to bite even experienced users and decrease the accuracy of their results with BM25, on the Gensim mailing list.

 [1]: https://pypi.org/project/rank-bm25/
 [2]: https://groups.google.com/g/gensim/c/wzQ1cIl9Img
 [3]: https://github.com/RaRe-Technologies/gensim/files/8200024/trec-rank-bm25.pdf
 [4]: https://github.com/RaRe-Technologies/gensim/files/8200025/trec-gensim.pdf

**Comment by piskvorky** [comment]
The `*MatrixSimilarity` stuff is the oldest part of Gensim, along with `LsiModel`. It dates back to DML-CZ days, in ancient pre-history :)

To me it makes perfect sense to control the normalization via a parameter. Are you able to add such option? We have to keep the defaults 100% backward compatible though.

**Comment by Witiko** [comment]
Not a problem, I added it to my task list. The `SoftCosineSimilarity` constructor uses a single parameter `normalized` that takes a tuple of two booleans, one for queries and one for documents, but let's deprecate that and have `normalize_queries` and `normalize_documents` across the `SimilarityABC` subclasses. Seems more readable.

**Comment by piskvorky** [comment]
OK.

**Comment by Witiko** [comment]
In https://github.com/RaRe-Technologies/gensim/commit/f43806d433955407c39f6f58a9045f847871b9d9, I halfway-implemented [BM25L][1], but I realized that it is difficult to fully implement as a sparse vector space model. That is because in BM25L, document vectors have the weight δ (typically set to δ = 0.5) for all terms that did not occur in the document, which eliminates sparsity. This could be implemented if `scipy.sparse` supported a flag that would make the value of *zero* elements not zero, but a different constant, which I doubt it does. Alternatively, we could have a special-purpose index just for BM25L, but that seems to defeat the purpose of implementing things in Gensim, which is interoperability with other vector space concepts and models. Therefore, I plan to abandon BM25L and focus at BM25<sup>+</sup> next.

[1]: http://www.cs.otago.ac.nz/homepages/andrew/papers/2014-2.pdf#page=2

**Comment by Witiko** [comment]
In be7a0e68, I halfway-implemented [BM25<sup>+</sup>](http://www.cs.otago.ac.nz/homepages/andrew/papers/2014-2.pdf#page=2), but similarly to BM25L, BM25<sup>+</sup> dictates that every query term contributes at least δ to the score even if it did not appear in the document, which makes it difficult to implement BM25<sup>+</sup> as a vector space model.

**Comment by piskvorky** [comment]
Isn't there a way to factor out such "constant" contributions?

Btw the sparse implementation in scipy is pretty "dumb" (in the sense of unoptimized). So writing a specialized for-loop or two with "non-zero" semantics shouldn't be terribly difficult, even without clever factoring math tricks.

**Comment by piskvorky** [comment]
Okay, up to you. I'm not familiar with the motivation BM25+ and BM25L etc, can't judge how relevant / in demand by users they are.

**Comment by Witiko** [comment]
They are quite relevant. For example, BM25<sup>+</sup> outperformed [[1](https://dl.acm.org/doi/10.1145/2063576.2063584)] Okapi BM25 on all TREC GOV2, WT100g, WT2G, and Robust04 information retrieval collections. However, the improvement is meager compared the improvement that Okapi BM25 brings over TF-IDF. Therefore, I would rather have Okapi BM25 now (should be ready for review by the end of the week) and add some of the more exotic variants of BM25 later on over inflating the scope of this PR and dragging it out for another month.

**Comment by Witiko** [comment]
@piskvorky The PR should be ready for review.

**Comment by Witiko** [comment]
I have performed a sanity check of https://github.com/RaRe-Technologies/gensim/pull/3304/commits/90bc47085e3527b9c8ab4d7d86203721320100a3 using [the TREC 6–8 collection][1]:

| Vector space model | MAP score |
|:--|:--|
| Bag of words | 6.00% |
| TF-IDF (nfc.nfc) | 13.68% |
| TF-IDF (dtb.nnn) | 22.83% |
| *OkapiBM25Model* | *24.93%* |
| **LuceneBM25Model** | **26.28%** |
| **AtireBM25Model** | **26.28%** |

I have already checked that our implementation of `OkapiBM25Model` agrees with `rank_bm25.OkapiBM25` in https://github.com/RaRe-Technologies/gensim/pull/3304#issuecomment-1061000689.

 [1]: https://github.com/RaRe-Technologies/gensim/files/8426218/Copy.of.Copy.of.Copy.of.Copy.of.Second.term.project.TREC.example.solution.ipynb.-.Colaboratory.pdf

**Comment by Witiko** [comment]
@piskvorky @gojomo Please, let me know if there is anything I can do to help move this PR along.

**Comment by gojomo** [comment]
I'm not close enough to prior related work, or potential future uses, of this to feel qualified to do a close review for correctness/risks/etc. My opinion is thus limited to: sounds useful, and a very quick glance over the diffs makes it look like docs/tests/clean-coding-practices are all in order. So I'd encourage promoting to others better-able to test/use.

**Comment by piskvorky** [comment]
@Witiko busy ATM, sorry. I plan to get to this in June – is there anything urgent riding on the merge from your side? Some deadlines?

**Comment by Witiko** [comment]
Ping @piskvorky and @gojomo. This pull request implements BM25, the only bag-of-words weighting scheme that is actually used for information retrieval in the wild. This is not a controversial statement: Lucene dropped TF-IDF in favor of BM25 [back in 2016][1], it's time to get Gensim back on track. The code is documented, unit-tested, benchmarked, passes the CI, and my review. Please, let me know if there is anything I can do to champion this PR and help move its review along.

 [1]: https://lists.apache.org/thread/33s93q2l32rgs5mvlg31mdnh7d56cokr

**Comment by ramsey-coding** [comment]
@piskvorky @gojomo when would Gensim provide support for BM25? This feature is hanging for a long long time now. 

Would this ever made it to the library!

**Comment by ramsey-coding** [comment]
@Witiko what's the equivalent api call for 

- `bm25.get_top_n(tokenized_query, corpus, n=80)`

**Comment by Witiko** [comment]
I don't think there is an equivalent. Get all similarities, get an argmax and take the top 80.

**Comment by dunefox** [comment]
This would have been very useful for me during the last few weeks. Sadly, there doesn't seem to be much interest in BM25 here.

**Comment by ramsey-coding** [comment]
> I don't think there is an equivalent APU call. Get all similarities, run argmax over them and take the top 80.

I don't follow it. You are saying for a given query, I iterate over the whole 350K datapoint and get similarity and then take top 80.

This would not scale at all. 😭

**Comment by ramsey-coding** [comment]
@Witiko also the API of gensim is neither user friendly nor convenienet. It appears it just provide the similarity score and does not even provide the original document.

It appears to me Gensim is an old, out-dated library and neither wants to move forward. Probably time to abandon this library and users should look for better alternatives.

**Comment by Witiko** [comment]
What I am saying is that it will be significantly faster than rank-bm25.

**Comment by ramsey-coding** [comment]
> What I am saying is that it will be significantly faster than rank-bm25.

got it

**Comment by Witiko** [comment]
Gensim will get you similarities in the order of indexing, i.e. if you index documents 1, 2, and 3, and then perform a query, you will get back similarities between the query and documents 1, 2, and 3, respectively.

**Comment by ramsey-coding** [comment]
@Witiko you are phenomenal, thanks for all the great feedback. 

I have one more question:

If I set `num_best=80` here:
```
SparseMatrixSimilarity(bm25_corpus,
                                        num_docs=len(corpus),
                                        num_terms=len(dictionary),
                                        normalize_queries=False,
                                        normalize_documents=False,
                                        num_best=80)
```

And then get similarity like the following, would the result would be sorted by most matched document?

```
    tfidf_model = TfidfModel(dictionary=bm25_dictionary, smartirs='bnn')  # Enforce binary weighting of queries
    tfidf_query = tfidf_model[bm25_dictionary.doc2bow(tokenized_query)]

    similarities = bm25_index[tfidf_query]
    for doc_no, score in bm25_index[tfidf_query]:
        print("original document:", test_methods_corpus[doc_no])
```

So question is here the result of `bm25_index[tfidf_query]` would be sorted based on most matched documents or not?

**Comment by Witiko** [comment]
Take `sorted(enumerate(bm25_index[tfidf_query]))[-1]` and you will get the index and the similarity of the most similar document to your query.

**Comment by smith-co** [comment]
@Witiko wow, awesome work. In the context of this implementation:

```
similarities = bm25_index[tfidf_query]
````

-  is `higher score` means the `document` is more `similar` to the query?
- Or `lower score` means the document is more similar to the query.

Sorry for the stupid question.

**Comment by nashid** [comment]
This feature would be very useful for me.

**Comment by Witiko** [comment]
@smith-co The similarities are BM25 scores, i.e. the higher the similarity, the more similar the document is to your query.

**Comment by Witiko** [comment]
@ramsey-coding @smith-co I added outputs to the example code in the original post and showed how you can get back the best document for a query. I hope you will find that useful.

**Comment by Witiko** [comment]
> So question is here the result of `bm25_index[tfidf_query]` would be sorted based on most matched documents or not?

@ramsey-coding Sorry, I wasn't at my computer over the weekend. Yes, your understanding is correct; specifying `num_best=80` in `SparseMatrixSimilarity(...)` will cause `bm25_index[tfidf_query]` to produce an iterable of 80 document ids and similarities sorted from the most matched document in the descending order of similarity.

**Comment by mgeletka** [comment]
I would really appreciate merging this functionality as I must now use my own custom implementation of BM25 when working with the Gensim library.

File: gensim/models/bm25model.py
Comment (piskvorky) [inline gensim/models/bm25model.py:164]: What does this line mean? It's all in caps and in emphasis, so I assume it's important. It it a part of the reference string, or some warning? What does `SP` stand for?

File: gensim/models/bm25model.py
Comment (piskvorky) [inline gensim/models/bm25model.py:390]: I'm surprised this wild indentation passed the linter :)

**Review by piskvorky — APPROVED** [review]

**Comment by piskvorky** [comment]
Code looks nice and clean, sorry for taking so long to review.

@mpenkov anything else we need before merge?

@Witiko how about post-merge? What can we do to promote this functionality (beyond including it in the Gensim gallery)?

File: gensim/models/bm25model.py
Comment (Witiko) [reply gensim/models/bm25model.py:164]: That's the name of the collection in which the article was published.

Comment (piskvorky) [reply gensim/models/bm25model.py:164]: OK I googled it up; it's simply this:

```suggestion
       *National Institute of Standard and Technology Special Publication (NIST SP)*.
```

**Comment by Witiko** [comment]
@piskvorky Thank you for taking the time. We can mention in the release notes that Gensim can now be used for Lucene-style information retrieval.

**Comment by piskvorky** [comment]
Sure, it goes into the release notes without saying.

I meant more like some demo, or a practical use-case (who'd use it and how?), or similar. A motivational part, to anchor the technical part.

Maybe @dunefox @mgeletka @smith-co @nashid @ramsey-coding could help?

**Comment by piskvorky** [comment]
@mpenkov anything missing here?

Let's aim to release soon after merging, to get this feature out. Thanks.

**Comment by mpenkov** [comment]
Sorry for the delay guys, merging.

**Comment by piskvorky** [comment]
Thanks MIsha!

@dunefox @mgeletka @smith-co @nashid @ramsey-coding could you write a few words about how you use Okapi BM25, or intend to use it?

Your story, your use-case, your motivation to participate in this PR.

**Comment by ditengm** [comment]
Hello @Witiko! 
Can you please tell me how to get word corpus embedding? That is, the corpus weight matrix?
Thanks!

**Comment by Witiko** [comment]
> Hello @Witiko! 
> Can you please tell me how to get word corpus embedding? That is, the corpus weight matrix?
> Thanks!

Hello, @ditengm. You can get the BM25 weight matrix of your corpus from `bm25_index.index`, where `bm25_index` is the `SparseMatrixSimilarity` index from the second example in the original post. The type of `bm25_index.index` is `scipy.sparse.csr_matrix`.

**Pull request merged**
